sdk/javascript/client/encoder.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2017 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+'use strict'
+
+const {createHash} = require('crypto')
+const Signer = require('./signer')
+const {
+  TransactionHeader,
+  Transaction,
+  BatchHeader,
+  Batch,
+  BatchList
+} = require('../protobuf')
+
+const _arrayify = itemOrItems => {
+  if (Array.isArray(itemOrItems)) return itemOrItems
+  return [itemOrItems]
+}
+
+const _makeNonce = () => {
+  const dateString = Date.now().toString(36).slice(-5)
+  const randomString = Math.floor(Math.random() * 46655).toString(36)
+  return dateString + ('00' + randomString).slice(-3)
+}
+
+/** Defines encoders which create multiple Transactions with common settings */
+class TransactionEncoder {
```

Comment:
Given that it deals with Transactions, Batches and Batch Lists, this name is too specific, making the batch and list methods a bit odd.  Either separate them into their own class, or give this a more generic name.